File: java/00309._best_time_to_buy_and_sell_stock_with_cooldown.java
//        Java
// Time complexity - O(n×2×2)=O(4n)=O(n)
// Space complexity - O (n)

public class Solution {
    public int maxProfit(int[] prices) {
        Map<String, Integer> memo = new HashMap<>();
        return generateTree(prices, false, false, 0, memo);
    }

    private int generateTree(int[] prices, boolean bought, boolean sold, int day, Map<String, Integer> memo) {
        if (day >= prices.length) {
            return 0;
        }

        String key = day + "_" + bought + "_" + sold;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }

        int maxProfit = 0;

        if (bought) {
            int sellProfit = generateTree(prices, false, true, day + 1, memo) + prices[day];
            int holdProfit = generateTree(prices, true, false, day + 1, memo);
            maxProfit = Math.max(sellProfit, holdProfit);
        } else {
            if (sold) {
                maxProfit = generateTree(prices, false, false, day + 1, memo);
            } else {
                int buyProfit = generateTree(prices, true, false, day + 1, memo) - prices[day];
                int skipProfit = generateTree(prices, false, false, day + 1, memo);
                maxProfit = Math.max(buyProfit, skipProfit);
            }
        }

        memo.put(key, maxProfit);
        return maxProfit;
    }
}

File: python/00309._best_time_to_buy_and_sell_stock_with_cooldown.py
#        Python
# Time complexity - O(n×2×2)=O(4n)=O(n)
# Space complexity - O (n)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        memo = {}
        return self.generate_tree(prices, False, False, 0, memo)

    def generate_tree(self, prices: List[int], bought: bool, sold: bool, day: int, memo: Dict[str, int]) -> int:
        if day >= len(prices):
            return 0

        key = f"{day}_{bought}_{sold}"
        if key in memo:
            return memo[key]

        max_profit = 0

        if bought:
            sell_profit = self.generate_tree(prices, False, True, day + 1, memo) + prices[day]
            hold_profit = self.generate_tree(prices, True, False, day + 1, memo)
            max_profit = max(sell_profit, hold_profit)
        else:
            if sold:
                max_profit = self.generate_tree(prices, False, False, day + 1, memo)
            else:
                buy_profit = self.generate_tree(prices, True, False, day + 1, memo) - prices[day]
                skip_profit = self.generate_tree(prices, False, False, day + 1, memo)
                max_profit = max(buy_profit, skip_profit)

        memo[key] = max_profit
        return max_profit
